File: 2_1.cpp
#include <iostream>
#include <random>
#include <vector>
#include <chrono>
#include <algorithm>
using namespace std;
vector <int> findld(const vector <int> & arr){ // втрое 
    int currentLength = 1;
    int maxLength = 1;
    int startIndex = 0;
    int maxStartindex = 0;
    int prev = arr.front();
    int index = 0;
    for (int num : arr) { // на первом этапе сравнивает два раза первое
        if (index > 0) {
            if (num < prev) {
                currentLength++;
            } else {
                if (currentLength > maxLength) {
                    maxLength = currentLength;
                    maxStartindex = startIndex;
                }
                currentLength = 1;
                startIndex = index;
            }
        }
        prev = num;
        index++;
    }
    if (currentLength > maxLength) {
        maxLength = currentLength;
        maxStartindex = startIndex;
    }
    vector<int> result;
    for (int i = maxStartindex; i < maxStartindex + maxLength; ++i) {
        result.push_back(arr[i]);
    }
    return result;
}
vector <int> ThirdVector (const vector <int> & arr){
    double srzn = 0;
    for (int n : arr){
        srzn += n;
    }
    srzn = srzn / arr.size();
    vector<int> result;
    for (double n : arr){
        if (srzn > n){
            result.push_back(n);
        }
    }
    return result;
}
vector <int> FourhVector (const vector <int> & arr){
    int first = arr[0], second = arr[1];
    int index = 0;
    int k = 0;
    vector<int> result;
    for (int n : arr){
        if (k < 2){
            k++;
            continue;
        }
        result.push_back(n);
    }
    result.push_back(first); result.push_back(second);
    return result;
}
vector <int> Fifth_1 (const vector <int> & arr){
    vector<int> result = arr;
    sort(result.begin(), result.end(), [](int a, int b) {
        return a > b; // Сортировка по убыванию: возвращаем true, если a должно идти перед b
    });
    return result;
};
vector<int> Fifth_2(const vector<int> &arr) {

    vector<int> digitCount(10, 0); // Массив для подсчёта количества вхождений каждой цифры от 0 до 9

    
    for (int num : arr) {
        int temp = num;
        while (temp > 0) {
            int digit = temp % 10; 
            digitCount[digit]++;   
            temp /= 10;             
        }
    }

    vector<int> result(digitCount.begin(), digitCount.end());//вхождений каждой цифры от 0 до 9

    return result;
}
int main () {
    vector <int> randomVector;
    vector <int> taskVector;
    unsigned seed = chrono::system_clock::now().time_since_epoch().count(); // чтобы были каждый раз разные используем текущее время
    ranlux48_base rng (seed); // ranlux генерирует по сидам и 
    int n = 10;
    for (int i = 0; i < n; i++) {
        randomVector.push_back(rng() % 201 + 150); // [150, 350]
    }
    taskVector = findld(randomVector);
    for (int n : randomVector){
        cout << n << " ";
    }
    cout << endl;
    for (int n : taskVector){
        cout << n << " ";
    }
    taskVector.clear();
    cout << endl;
    taskVector = ThirdVector(randomVector);
    for (int n : taskVector){
        cout << n << " ";
    }
    taskVector.clear();
    cout << endl;
    randomVector.clear(); // поменяем диапозон рандома
    for (int i = 0; i < n; i++) {
        randomVector.push_back(rng() % 93 + 31); // по ascii таблице
    }
    for (char n : randomVector){ // просто неявно преабразуем 
        cout << n << " ";        
    }
    cout << endl;
    taskVector = FourhVector(randomVector);
    for (char n : taskVector){
        cout << n << " ";
    }
    taskVector.clear();
    cout << endl;
    randomVector.clear();
    for (int i = 0; i < n; i++) {
        randomVector.push_back(rng() % 801 + 100); // [100,900]
    }
    taskVector = Fifth_1(randomVector);
    for (int n : taskVector){
        cout << n << " ";
    }
    cout << endl;
    taskVector.clear();
    taskVector = Fifth_2(randomVector);
    for (int n : taskVector){
        cout  << n << " ";
    }
}

File: go/2_1.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
	"time"
)

func findld(arr []int) []int {
	currentLength := 1
	maxLength := 1
	startIndex := 0
	maxStartIndex := 0
	prev := arr[0]

	for index, num := range arr {
		if index > 0 {
			if num < prev {
				currentLength++
			} else {
				if currentLength > maxLength {
					maxLength = currentLength
					maxStartIndex = startIndex
				}
				currentLength = 1
				startIndex = index
			}
		}
		prev = num
	}

	if currentLength > maxLength {
		maxLength = currentLength
		maxStartIndex = startIndex
	}

	result := make([]int, maxLength)
	copy(result, arr[maxStartIndex:maxStartIndex+maxLength])
	return result
}

func thirdVector(arr []int) []int {
	sum := 0
	for _, n := range arr {
		sum += n
	}
	avg := float64(sum) / float64(len(arr))

	var result []int
	for _, n := range arr {
		if float64(n) < avg {
			result = append(result, n)
		}
	}
	return result
}

func fourthVector(arr []int) []int {
	first, second := arr[0], arr[1]
	result := append([]int{}, arr[2:]...)
	result = append(result, first, second)
	return result
}

func fifth1(arr []int) []int {
	result := make([]int, len(arr))
	copy(result, arr)
	sort.Sort(sort.Reverse(sort.IntSlice(result)))
	return result
}

func fifth2(arr []int) []int {
	digitCount := make([]int, 10)

	for _, num := range arr {
		temp := num
		for temp > 0 {
			digit := temp % 10
			digitCount[digit]++
			temp /= 10
		}
	}

	return digitCount
}

func main() {
	rand.Seed(time.Now().UnixNano())

	var randomVector []int
	n := 10

	// Заполняем срез случайными значениями от 150 до 350
	for i := 0; i < n; i++ {
		randomVector = append(randomVector, rand.Intn(201)+150) // [150, 350]
	}

	taskVector := findld(randomVector)
	fmt.Println("Original Random Vector:", randomVector)
	fmt.Println("Result of findld:", taskVector)

	randomVector = nil
	// Заполняем срез случайными значениями ASCII (от 31 до 123)
	for i := 0; i < n; i++ {
		randomVector = append(randomVector, rand.Intn(93)+31) // ASCII range (31-123)
	}

	fmt.Println("Random ASCII Vector:")
	for _, num := range randomVector {
		character := rune(num) // Преобразуем int в Unicode символ
		fmt.Printf("%c ", character)
	}
	fmt.Println()

	taskVector = fourthVector(randomVector)
	for _, num := range taskVector {
		character := rune(num) // Преобразуем int в Unicode символ
		fmt.Printf("%c ", character)
	}
	fmt.Println()
	randomVector = nil
	// Заполняем срез случайными значениями от 100 до 900
	for i := 0; i < n; i++ {
		randomVector = append(randomVector, rand.Intn(801)+100) // [100, 900]
	}

	taskVector = fifth1(randomVector)
	fmt.Println("Result of fifth1 (sorted descending):", taskVector)

	taskVector = fifth2(randomVector)
	fmt.Println("Result of fifth2 (digit count):", taskVector)
}
